### PR TITLE
Improves the `reload_lib -a` commands ability to track modified files

### PR DESCRIPTION
**Pull request by cgranleese-r7**
The reload_lib functionality works based on git diff, for example: if you modify a file with a breakpoint or similar, 
`reload_lib -a`, then remove the breakpoint - further `reload_lib -a` calls won't remove the breakpoint as it's no longer part of the git diff.

This PR adds support for tracking the previously reloaded files, so that future reload_lib calls load these files again in the future.

### Scenario 1 -  ensure changed files are being tracked on boot
- [x] Add a `pry` to the [search command](https://github.com/rapid7/metasploit-framework/blob/aafb38c7ed37c6c7cf8f2bfbf2cc9edb478b9123/lib/msf/ui/console/command_dispatcher/modules.rb#L397)
- [x] Start `msfconsole`
- [x] Run a search command e.g. `search eternalblue`
- [x] **Verify** you hit the breakpoint and run `continue` to get back the console
- [x] Remove the search command breakpoint and run `reload_lib -a`
- [x] Redo the search command and verify you no longer trigger the breakpoint

### Scenario 2 - ensure file are being tracked when modified whilst `msfconsole` is running.
- [x] Start `msfconsole`
- [x] Add a `pry` to the [search command](https://github.com/rapid7/metasploit-framework/blob/aafb38c7ed37c6c7cf8f2bfbf2cc9edb478b9123/lib/msf/ui/console/command_dispatcher/modules.rb#L397)
- [x] Run `reload_lib -a`
- [x] Run a search command e.g. `search eternalblue`
- [x] **Verify** you hit the breakpoint and run `continue` to get back the console
- [x] Remove the search command breakpoint and run `reload_lib -a`
- [x] Redo the search command and verify you no longer trigger the breakpoint

### CODE REVIEW

File: lib/msf/ui/console/command_dispatcher/developer.rb [hunk 1]
```diff
@@ -435,4 +433,16 @@ def cmd_time_help
     print @@time_opts.usage
     print_line
   end
+
+  private
+
+  def modified_files
+    # Using an array avoids shelling out, so we avoid escaping/quoting
+    changed_files = %w[git diff --name-only]
```
Comment (gwillcox-r7) [inline lib/msf/ui/console/command_dispatcher/developer.rb:441]: Um isn't this meant to be like this? 
```suggestion
    changed_files = %w(git diff --name-only)
```

Though this avoided the array syntax.

**Review by gwillcox-r7 — APPROVED** [review]

**Comment by gwillcox-r7** [comment]
Before:

```
msf6 > search eternalblue

From: /home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/modules.rb:412 Msf::Ui::Console::CommandDispatcher::Modules#cmd_search:

    407:             valid_sort_attributes = ['rank','disclosure_date','name','date','type','check']
    408:             reverse_sort = false
    409:             ignore_use_exact_match = false
    410: 
    411:             require 'pry'; binding.pry
 => 412:             @@search_opts.parse(args) do |opt, idx, val|
    413:               case opt
    414:               when '-S'
[1] pry(#<Msf::Ui::Console::CommandDispatcher::Modules>)> continue

Matching Modules
================

   #  Name                                      Disclosure Date  Rank     Check  Description
   -  ----                                      ---------------  ----     -----  -----------
   0  exploit/windows/smb/ms17_010_eternalblue  2017-03-14       average  Yes    MS17-010 EternalBlue SMB Remote Windows Kernel Pool Corruption
   1  exploit/windows/smb/ms17_010_psexec       2017-03-14       normal   Yes    MS17-010 EternalRomance/EternalSynergy/EternalChampion SMB Remote Windows Code Execution
   2  auxiliary/admin/smb/ms17_010_command      2017-03-14       normal   No     MS17-010 EternalRomance/EternalSynergy/EternalChampion SMB Remote Windows Command Execution
   3  auxiliary/scanner/smb/smb_ms17_010                         normal   No     MS17-010 SMB RCE Detection
   4  exploit/windows/smb/smb_doublepulsar_rce  2017-04-14       great    Yes    SMB DOUBLEPULSAR Remote Code Execution


Interact with a module by name or index. For example info 4, use 4 or use exploit/windows/smb/smb_doublepulsar_rce

msf6 > reload_lib -a
msf6 > search eternalblue

From: /home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/modules.rb:412 Msf::Ui::Console::CommandDispatcher::Modules#cmd_search:

    407:             valid_sort_attributes = ['rank','disclosure_date','name','date','type','check']
    408:             reverse_sort = false
    409:             ignore_use_exact_match = false
    410: 
    411:             @@search_opts.parse(args) do |opt, idx, val|
 => 412:               case opt
    413:               when '-S'
    414:                 row_filter = val
```

**Comment by cgranleese-r7** [comment]
Sorry @gwillcox-r7 I double checked this this morning and realised I missed a verification step to run `reload_lib -a` in scenario 2, my bad. I have updated the verification steps to add the missing step.

**Comment by gwillcox-r7** [comment]
Looks like updated second scenario is now working:

```
msf6 > reload_lib -a 
[*] Reloading /home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/modules.rb
msf6 > search externalblue

From: /home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/modules.rb:411 Msf::Ui::Console::CommandDispatcher::Modules#cmd_search:

    406:             valid_sort_attributes = ['rank','disclosure_date','name','date','type','check']
    407:             reverse_sort = false
    408:             ignore_use_exact_match = false
    409: 
    410:             require 'pry'; binding.pry
 => 411:             @@search_opts.parse(args) do |opt, idx, val|
    412:               case opt
    413:               when '-S'
[1] pry(#<Msf::Ui::Console::CommandDispatcher::Modules>)> continue
[-] No results from search
msf6 > externalblue
[-] Unknown command: externalblue
msf6 > search eternalblue

From: /home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/modules.rb:411 Msf::Ui::Console::CommandDispatcher::Modules#cmd_search:

    406:             valid_sort_attributes = ['rank','disclosure_date','name','date','type','check']
    407:             reverse_sort = false
    408:             ignore_use_exact_match = false
    409: 
    410:             require 'pry'; binding.pry
 => 411:             @@search_opts.parse(args) do |opt, idx, val|
    412:               case opt
    413:               when '-S'
    414:                 row_filter = val
[1] pry(#<Msf::Ui::Console::CommandDispatcher::Modules>)> continue

Matching Modules
================

   #  Name                                      Disclosure Date  Rank     Check  Description
   -  ----                                      ---------------  ----     -----  -----------
   0  exploit/windows/smb/ms17_010_eternalblue  2017-03-14       average  Yes    MS17-010 EternalBlue SMB Remote Windows Kernel Pool Corruption
   1  exploit/windows/smb/ms17_010_psexec       2017-03-14       normal   Yes    MS17-010 EternalRomance/EternalSynergy/EternalChampion SMB Remote Windows Code Execution
   2  auxiliary/admin/smb/ms17_010_command      2017-03-14       normal   No     MS17-010 EternalRomance/EternalSynergy/EternalChampion SMB Remote Windows Command Execution
   3  auxiliary/scanner/smb/smb_ms17_010                         normal   No     MS17-010 SMB RCE Detection
   4  exploit/windows/smb/smb_doublepulsar_rce  2017-04-14       great    Yes    SMB DOUBLEPULSAR Remote Code Execution


Interact with a module by name or index. For example info 4, use 4 or use exploit/windows/smb/smb_doublepulsar_rce

msf6 > removing breakpoint :)
[-] Unknown command: removing
msf6 > reload_lib -a
[*] Reloading /home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/modules.rb
msf6 > 
msf6 > search eternalblue

Matching Modules
================

   #  Name                                      Disclosure Date  Rank     Check  Description
   -  ----                                      ---------------  ----     -----  -----------
   0  exploit/windows/smb/ms17_010_eternalblue  2017-03-14       average  Yes    MS17-010 EternalBlue SMB Remote Windows Kernel Pool Corruption
   1  exploit/windows/smb/ms17_010_psexec       2017-03-14       normal   Yes    MS17-010 EternalRomance/EternalSynergy/EternalChampion SMB Remote Windows Code Execution
   2  auxiliary/admin/smb/ms17_010_command      2017-03-14       normal   No     MS17-010 EternalRomance/EternalSynergy/EternalChampion SMB Remote Windows Command Execution
   3  auxiliary/scanner/smb/smb_ms17_010                         normal   No     MS17-010 SMB RCE Detection
   4  exploit/windows/smb/smb_doublepulsar_rce  2017-04-14       great    Yes    SMB DOUBLEPULSAR Remote Code Execution


Interact with a module by name or index. For example info 4, use 4 or use exploit/windows/smb/smb_doublepulsar_rce

msf6 > 
```

**Comment by gwillcox-r7** [comment]
LTGM will land this now :)

**Comment by gwillcox-r7** [comment]
# Release Notes
The `reload_lib` functionality has been updated so that its file change tracking logic better takes into account scenarios where files are modified. Previously if a breakpoint was inserted, removed, and then `reload_lib -a` was run, it would mistakenly use an old copy of the code.

**Pull request merged**